### ISSUE TITLE
wrongs of goal changing

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#612]
Situation: I've changed goal from _xor2_ to _ec41_ on folder about 35 TiB in size. At the moment of goal change cluster had around 18 TiB of free space. Now were are facing the following problems:

  * We are rapidly running out of space. More than 12 TiB is already consumed by incomplete goal change (that should have released some space). Why goal can't be changed transactionally one chunk at a time?

  * New goal is defined as `$ec(4,1){ poolA poolA poolB poolB wks }`. There are at least two chunkservers in _poolA_ and _poolB_ and those chunkservers are always up. There are 5 chunkservers in pool _wks_ and we get unavailable files when some of _wks_ chunkserrvers are stopped. New goal was specifically designed to prevent that situation so some chunks must be in degraded state while their goal is being changed (this can't be safe).

 * As goal change is threatening to consume all available space I'm trying to pin some unimportant files  to temporary chunkserver labelled "_bad_" by creating only one goal that put one and only replica to new chunkserver. As soon as "_bad_" chunkserver is started it began to receive traffic despite the fact that there were still no files with goal "_bad_" and no other goals defined with '_' so that chunkserver should not have been used even indirectly. Quick glimpse at the "_bad_" file system revealed chunks like `chunk_ec_1_of_4_1_` so goal change places chunks where they do not belong in violation of explicit goal definition. Incorrect placement surely does not help to change goal faster due to needless/excessive replication traffic...

I'm surprised to find how bad a goal change can be...

Summary:
  * Excessive space consumed by goal change.
  * Incorrect chunks placement (goal violation).

**Comment by psarna** [comment]
In your specific case, for a single chunk, no xor copies would be removed until all ec(4,1) copies are created, because only then needed safety level is reached. More details below.

To address your question on transactional goal change - the short answer is that's not how it works in LizardFS. We have a chunk loop which goes through chunks periodically and does 1 thing at a time: replication, deletion, rebalance, etc. Due to this design, it's way more likely that 400 chunks would have 1 ec part created during goal change than 100 chunks having all their parts created at once. It's done that way to ensure that no chunks are starved - otherwise huge goal change could cause other chunks not to be processed for a long time. True, huge goal changes between xor/ec pretty much need space for both goals to fit in your installation. You can make it more transactional by changing goals successively, e.g. only a couple directories at a time.

Traffic occurs on chunkservers with `bad` label most likely because replication was needed and all other chunkserver hit their limits. Changing goal, especially a xor/ec one, is tricky, so master will always prefer to have excessive copies for a while (hence, space consumed), than risk losing data.
If master server decides that a chunk is endangered (and now it also depends on your `REDUNDANCY_LEVEL` configuration option) and all the other chunkservers are past their limits (`CHUNKS_WRITE_REP_LIMIT`, `CHUNKS_READ_REP_LIMIT`, `CHUNKS_SOFT_DEL_LIMIT`, etc.) it will perform cross-label replication rather than keeping this particular chunk endangered. From master's point of view, it's better to violate goal label than keep not enough copies of data.

Also, if your `ENDANGERED_CHUNKS_PRIORITY` is set to something above zero, then it's more likely that each chunk loop will take up more space, since these are the ones that need replication. Only after some of these endangered chunks are served, deletion of unneeded part can occur. And, because your goals are xor2 and ec41, no parts of xor2 would be removed before all ec41 are created to keep all chunks from being endangered.

**Comment by onlyjob** [comment]
Thanks for explanation. Indeed there is room for improvements. It is good to have excessive copies but it would be nice to __avoid processing all chunks simultaneously__. This is the only way to avoid consuming excessive space and bandwidth during goal change.

Also it would be great to __prevent goal spills__ where chunks are replicated in violation of their goals. This is problematic as goal change consumed _all_ free space in my cluster leaving less than 1% even in most important pools/chunkservers. Currently there is no way to prevent this bad and very undesirable behavior. Apparently goal spills made the whole situation worse by replicating essential chunks to unstable (less reliable) chunkservers...

P.S. I could not change goal for "couple directories at a time" as most of the space is consumed by few very large files in one directory (though I could change goal for one file at a time but now it's too late)...

**Comment by onlyjob** [comment]
Looks like setting `ENDANGERED_CHUNKS_PRIORITY` to `0` (default) helped with free space problem a little...

**Comment by psarna** [comment]
It should, by setting it to zero you allow more deletions of excessive chunks. Still, this big goal change operation will probably continue to demand space. There's one more way to slow it down, you could create another xor(2, 1) goal and set goal of some of your files back to xor(2, 1). Preferably, you should change goal back only in files that haven't created a lot of ec parts yet. Then, after the rest is properly moved to ec goals, change goals of these files again. I'm not sure if it still makes sense, because it would also trigger deleting some of already created ec parts, so it will extend the whole operation in time even more. On the other hand, that way you can limit the space needed by goal change.

**Comment by onlyjob** [comment]
Thanks. Idea of reverting goal back to original one for some files already crossed my mind but I'm not sure if doing it still makes sense half way through... :)

**Comment by onlyjob** [comment]
Another thing that bothers me is growth of _undergoal_ and _endangered_ chunks in unrleated goals.
Changing goal should be low priority but in reality massive goal change creates enough pressure to delay replication of (other) undergoal chunks due to large queue of chunks to be replicated...
Perhaps goal change replication queue should be segregated?
Changing goal is currently unsafe because it delays replication of undergoal data (that has nothing to do with goal change)...

**Comment by onlyjob** [comment]
I could not recover from this bad goal change without deleting massive amount of data. Even reverting goal did not help because all available space on _all_ chunkservers is already consumed. Adding more HDDs gave little relief as they were quickly filled then replication slowed down... The whole performance of cluster plummeted ever since goal change because of free space exhaution...
I even tried to create new chunkserver with new label and pin some data to it (to reduce scope of the goal change) but for whatever reason that did not yield desirable effect because _ec41_ chunks quickly filled all available storage in violation of their goal...

**Comment by onlyjob** [comment]
I feel that something is wrong with deletion strategy as I have far too many chunks "which need deletion": about 5_000_000+ and tens of thousands with 10+ copies to delete... Amount of free space consumed by goal change is massively excessive...

**Comment by psarna** [comment]
Whoah, 10+ copies is definitely weird. So, to be sure, this is 3.11.3 on Debian, goal change of a big directory from xor2 to ec41. I'll try to reproduce this, but we never encountered more copies to delete than the number actually needed by both goals...

**Comment by psarna** [comment]
Could you paste some of your `lizardfs fileinfo` results? Preferably the most unusual ones you see... that could help pinpoint the issue.

**Comment by onlyjob** [comment]
Yes, 3.11.3 (latest) on Debian (official package from native repository, not a vendor packages).

Here is a fragment of `fileinfo`:

~~~~
        chunk 312: 00000D9CA9BB5CF5_00000001 / (id:14966513687797 ver:1)
                copy 1: 192.168.0.2:9422:pool part 3/5 of ec(4,1)
                copy 2: 192.168.0.3:9622:pool part 2/3 of xor2
                copy 3: 192.168.0.3:9622:pool part 5/5 of ec(4,1)
                copy 4: 192.168.0.4:9422:wks part 2/5 of ec(4,1)
                copy 5: 192.168.0.7:9422:wks part 5/5 of ec(4,1)
                copy 6: 192.168.0.111:9422:wdia part 3/5 of ec(4,1)
                copy 7: 192.168.0.136:9422:wks part 4/5 of ec(4,1)
                copy 8: 192.168.0.204:9422:wks part 3/3 of xor2
                copy 9: 192.168.0.250:9522:stor part 1/3 of xor2
                copy 10: 192.168.0.250:9522:stor part 1/5 of ec(4,1)
...
        chunk 319: 00000D9CA9BB69FB_00000001 / (id:14966513691131 ver:1)
                copy 1: 192.168.0.2:9422:pool part 2/3 of xor2
                copy 2: 192.168.0.2:9422:pool part 1/5 of ec(4,1)
                copy 3: 192.168.0.3:9622:pool part 1/5 of ec(4,1)
                copy 4: 192.168.0.4:9422:wks part 2/5 of ec(4,1)
                copy 5: 192.168.0.7:9422:wks part 3/5 of ec(4,1)
                copy 6: 192.168.0.204:9422:wks part 3/3 of xor2
                copy 7: 192.168.0.204:9422:wks part 5/5 of ec(4,1)
                copy 8: 192.168.0.204:9622:bad part 4/5 of ec(4,1)
                copy 9: 192.168.0.250:9422:stor part 4/5 of ec(4,1)
                copy 10: 192.168.0.250:9522:stor part 2/5 of ec(4,1)
~~~~

**Comment by psarna** [comment]
Multiple same parts of ec are troubling... Do these files have snapshots? Maybe for some reason they have a multiple goal definitions for single chunk.
The situation I'm talking about is something like:
file1 - goal ec(4,1){A A A _ _}
file2 - snapshot of file1 - goal ec(4,1){B B C C D}

In this case, if, let's say, both files share chunk number 00000000007, master would create more than 5 ec(4,1) parts of this chunk to fulfil all goal definitions and labels.

**Comment by onlyjob** [comment]
No snapshots... As far as I can tell this is straightforward scenario but too many chunks may be because cluster ran out of space during goal change... As you've said earlier, pinning to labels does not work when there is not enough space... Though I can not explain delayed chunks removal...

**Comment by gutleib** [comment]
Can confirm that goal affinity to certain servers is not honoured.
Given: a 70% full filesystem with replica 2 goal. 2 types of servers h for hdd, s for ssd. S were empty, as no goals include s target.
The goal change (to ec goal) was applied to 12% of dataset or so. Now, the h servres are 75% full, s servers are 15% full.
No disks are full, no snapshots.
No ideas.

**Comment by onlyjob** [comment]
In my case what seems to trigger the problem is when one of chunkservers labelled `stor` became full so there were nowhere to save new chunks as per defined goal. Those chunks that should have been _only_ made on a particuar chunkserver spilled to (all?) other chunkservers causing massive performance and (lack of) free space problems.

Interestingly enough `stor` chunkservers were filling unevenly: at the moment of goal change two `stor` chunkservers had similar amount of free space (whilst different total capacity). Smaller chunkserver filled faster and that triggered further troubles. 

On top of that I tried to recover by re-naming one of chunkservers and creating new goal `1` pinned to that chunkserver. That only made situation worse as instead of reducing number of chunks it created more chunks on all chunkservers in blatant violation of goal. That made 10th copy as far as I understand.

IMHO first problem is unrestricted replication and goals violation. **When there is no place for a chunk it should not be made anywhere else but on designated chunkservers as per goal definition.**
Second problem appears to be uneven distribution of data that causes some chunkservers of the same label to fill faster.
When one or more chunkservers are filled up to 99% situation becomes really bad because LizardFS still tries to place chunks to full chunkserver instead of preferring chunkserver(s) with more available free space.

I'm surprised how difficult it might be to operate nearly full LizardFS cluster with less than trivial goals...

**Comment by psarna** [comment]
To address the first problem (unrestricted replication and goals violation): what we have here is a consensus between replication limits, labels and safety. 
When a chunk needs to be replicated, we have to pick a place for it. Let's say its goal is 2 copies on label A, but all chunkservers with label A hit their replication limits already. Three options to choose from now:
1. Don't replicate - violate file safety
2. Replicate to A anyway - violate replication limits
3. Replicate somewhere else - violate labeling

All of the above have pros and cons, and we just picked 3. I think that whatever we choose, somebody could object. With 1, someone could say that safety should go first and it's better to replicate anywhere than nowhere at all. With 2, people could argue that there's no point in replication limits at all if they're not taken into account.

I'm open to discussion about these heuristics though, would you rather pick another option (or, even better, do you have yet another approach in mind)? The biggest argument against 1 is that it would limit replication in case of a serious breakdown, when, let's say, all chunkservers with label A die in a fire.

**Comment by gutleib** [comment]
Thank you for clearing the reasoning behind such mechanics.
I would object, though, that this is optimal.
1) Say, you have different segments of single FS. Archives, important data and ephemeral data. Now one of your server with archives (slow and large) goes down. Smaller servers with important data and fast ones with ephemeral slowly overfill and the whole system becomes unusable, and, what is more troubling - won't recover unless you add storage >= than the lost one. Basicaly, you can't manually stop the emergency - you have lesser priority than the paranoid logics of the system. I say, this is an issue.
2) Now, difference between chunk state "missing 1 copy" with Replica3 (and higher), EC4,2 and Replica2 or Xor3 is vast. Should endangered chunks overspill to different labeled servers? Maybe. Should 5-th replica, or even 3-rd do the same, given the risk for other servers? I think it's not optimal.
3) But than again, that's my opinion. Why can't this be configurable? A MFSALLOWLABELSPILL=NO|ENDANGERED|YES in mfsmaster.cfg would be a perfect solution. There are different workflows and different sysops... Some would panic over unpredicted behaviour, some consider monitoring annoying.

And regarding our case - well... SSD servers are clearing, but the HDD never went over 76% full. I think, servers were used to spread load of rebuilding evenly. I think this is also a violation of configuration, which is annoying at least. Could this be configurable too?

**Comment by psarna** [comment]
Most of things can become configurable, it's just the matter of resources (in this particular case - programmers) ;) But generally you're right, something similar to ALLOWLABELSPILL=ENDANGERED is a more balanced solution. I'll mark this issue [enhancement] and go back to it after we deal with more urgent tasks.

**Comment by onlyjob** [comment]
I strongly _strongly_ suggest `ALLOWLABELSPILL=NEVER`. Violating goals is never useful and (as I've learned the hard way) it is quite harmful.

Running out of space in a particular goal is no drama -- admin can always redefine goal, change goal or add more space to cluster...

Violating goals is unsafe for other goals and for performance of cluster. Goals can be used to segregate data and pin it to a particular hardware. Violating goals is disastrous in such case.

If there is not enough space for a goal then safety is already threatened. Replicating in goals violation might exhaust space somewhere else and amplify the damage causing even more troubles with safety and domino-style failures all over the cluster.

When one goal is running out of space replication _must_ stop or it will threaten safety of data in other goals. Moreover when replication stops admin can easily identify reason (lack of space) while unrestricted replication makes it more difficult to see the problem.

Running out of space in one goal is better than running out of space everywhere (which is precisely what happened in my cluster).

IMHO this is outstanding problem and I hope it can be prioritised soon.

**Comment by guestisp** [comment]
So are you suggesting to keep endangered chunks, thus violating the goal and keeping data unprotected, just to not violating label?

If you have only one chunk where you should have 3 copies, your data is at risk. I this that in this case lizard should do anything possible to guarantee the redundancy fixing the endangered chunks (undergoal is far better than endangered)

**Comment by 4Dolio** [comment]
I split my data beyond goals into isolated clusters, however many share the same underlying disk storage..  i must yield to others experiences with label violations, but it does not sound good.

As for how to handle such situations.. I would hope that Endangered(1 copy) should be given a higher priority than undergoal(2 copy but not @goal >2).  I might go further and say it would be preferable to sacrifice existing safe(>2) copies on purpose and make them undergoal if that is needed to resolve existing endangered (1) copies due to lack of space... Before resorting to violating labels which crossed storage boundaries.

**Comment by onlyjob** [comment]
In LizardFS goals are defined using labels therefore labels are sacred. If labels are violated then goals are invalid then data placement is unreliable (unpredictable). It is imperative to avoid spills between labels or free space may be consumed everywhere (not just on some chunkservers). It is cluster admin's responsibility to ensure that labels are defined properly and that there is enough space for LizardFS to create all required chunks, as per goal definition, on (and within) designated chunkservers.

Spilling chunks onto other chunkservers causes excessive traffic and danger of cluster wide domino effect failures on all chunkservers.

When space is limited there may be not enough room to create a chunk. This problem can not be  fixed by violating goals and labels without severe side effects. Lack of free space in any part of the cluster is bad enough on its own but at least solution to lack of free space is somewhat obvious and straightforward. Lack of space should not propagate and spread uncontrollably. One of the reasons for goals is to compartmentalise and prevent "outbreaks" of uncontrolled replication.

**Comment by gutleib** [comment]
Well, I have percise the same point ad @onlyjob, but, thankfully I did not have such a problem myself.
This is exactly why I suggest to make this configurable and default to ENDANGERED. Responsible admins would change it to NO (NEVER), the Next>Next>Complete! crowd would be more comfortable with defaults.
Also, I would like to raise 2 issues that were overlooked:
1) Goal overspill during goal change without space problem. Why!? There is no crisis, why completely uninvolved labels are used? Yes, they are emptied afterwards, but network load and unwelcome usage is happening anyway.
2) The overall goal changing process logic. Instead of taking care of a single chunk at a time, making its new goal chunks stable and then deleting the old one, it creates a wave of new partial chunks, then another one and continues so, until the goal is reached and old chunk can be deleted. This is actually what leads to visibility of the whole "overspill" issue, and creates unexpected consequences for a unaware admin.

I understand, that this is a really big change, but I think the rehaul of this mechanism can be considered. This is a almost-HSM functionality, and it can be a strong feature, not an additional risk, as it is now. Is anyone interested in making this a separate enhancement? I think I could write a desired workflow/options from user and business standpoints.

**Comment by 4Dolio** [comment]
There is also the perhaps tangentially related inability to replicate to self as a CS between disks.  Perhaps this is coming into play while replicating between complex goal types?  Without the ability to self-replicate we have to push chunks over the wire, perhaps to an undesired label, and then bring it back to the correct label, which may have been on the same CS at which the chunk originally was.  Since a CS can not simply read old goal from self and refactor into new goal of a whole different encoding methods.  Perhaps it would be possible to include localhost in the CS source list when reading for the purpose of replication?  No harm if localhost has no copies to provide, but if it does It should naturally attempt that first?  Perhaps this is causing a sort of label/encoding/local vs network race.  Perhaps I am describing what I believe could be happening and it is not correct.  It feels like perhaps CS self evaluation could greatly speed up and simplify some operations by pseudo eliminating networking overhead by utilizing localhost (similar to how clients inherently prefer self as first attempt CS if available).

**Comment by Blackpaw** [comment]
>Violating goals is unsafe for other goals and for performance of cluster. Goals can be used to segregate data and pin it to a particular hardware. Violating goals is disastrous in such case.

Tend to agree with this - if the system is running out of space for a set of goals I'd rather resolve it manually then rely on automation. Far to easy as you say for the system to cascade out of control.

I can setup monitoring and warning procedures (zabbix etc) to let me know when problems occur.

**Comment by psarna** [comment]
I'll take a look at how to improve this mechanism so label spilling can be avoided, perhaps as a configurable option. It may take a while though, as I already have a huge priority queue of things to do and it's not getting any shorter.

**Comment by psarna** [comment]
OK, just a quick update. I browsed the code to find a good place for implementing the solution, and it turns out it should already work:

1. According to the code, if any chunkserver that matches a label exists, but it just hit its replication limits, **no label spill should occur**. Instead, master will delay the replication after limits are refreshed.
2. Ergo, there's **either** a bug somewhere which allows chunks to spill over labels **or** `CHUNKS_REBALANCING_BETWEEN_LABELS` is set in your master config (is it? Please confirm).
3. If there's a bug in here, please give me as much more context as possible: it would be great to get a `fileinfo` output of some example overreplicated file again, but accompanied with `getgoal` and your exact goal definition for that goal. The more complete info I get about this one overreplicated file, the sooner I can figure out what's wrong.

P.S. It also means that one of my previous posts is false - we actually delay replication for a while if there are matching labels on the installation, but replication limits were hit. So it's closer to option 1 from my previous post than to option 3.

**Comment by psarna** [comment]
And, of course, there is one more option to choose from in point 2. of my previous post. If you actually have no more available* chunkservers from given label, then that's why the spill happens.

*to be precise, chunkserver is considered as "not available" for replication if it's disconnected **or** it already has another copy of this particular chunk. Then, current politics of replication is to ignore labels and replicate anywhere (and this politics can be amended to replicate only if chunk is endangered, that's generally a good idea and I'll probably code it soon).

**Comment by psarna** [comment]
Now that I wrote it, it makes more sense - the rest of your chunkservers with given label were probably full due to massive goal change. That's the reason of your spill. So, your problem would for 99% be solved by reimplementing replication so it only spills through labels if given chunk is endangered.

**Comment by psarna** [comment]
For alpha testing, if anyone is interested: http://cr.skytechnology.pl:8081/#/c/3174/
(If anyone was, please post feedback here)

**Comment by onlyjob** [comment]
>  chunkserver is considered as "not available" for replication if it's disconnected, full or it __already has another copy of this particular chunk__.

I think that's my case. (Of course `CHUNKS_REBALANCING_BETWEEN_LABELS` is not set.)
LizzardFS should be able to change goal in-place without violating goals/labels. If goal demands that certain data should sit on designated chunkservers then even goal change should not move data away from chunkserver where it belongs.

**Comment by onlyjob** [comment]
I'm testing c3174 and so far it looks OK. I couldn't reproduce cross-label replication on goal change yet though I did not try too hard... :)

**Comment by onlyjob** [comment]
I've completed another massive goal change with c3174 patch applied and I did not see any cross label spills. Very good. I think replication finished sooner as well thanks to obeying goals boundaries.

Strange thing though is uneven chunkservers utilisation during goal change: when there are two chunkservers with same label and utilisation percentage, one chunkserver was running out of space by the end of goal change. This triggered re-balance that could have been avoided...

I'm all for merging this commit -- it is very useful. Thanks.
